// react-cosmos-loader/dist/components/Loader/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime-corejs2/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Loader = void 0;

var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/toConsumableArray"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = require("prop-types");

var _linkedList = _interopRequireDefault(require("@skidding/linked-list"));

var _moduleType = require("../../utils/module-type");

var _PropsProxy = require("../PropsProxy");

/**
 * Loader for rendering React components in isolation.
 *
 * Renders components using fixtures and Proxy middleware. Initialized via
 * props.
 */
var Loader = function Loader(_ref) {
  var fixture = _ref.fixture,
      proxies = _ref.proxies,
      onComponentRef = _ref.onComponentRef,
      onFixtureUpdate = _ref.onFixtureUpdate;
  var firstProxy = (0, _linkedList.default)((0, _toConsumableArray2.default)(proxies).concat([_PropsProxy.PropsProxy]));
  return _react.default.createElement(firstProxy.value, {
    nextProxy: firstProxy.next(),
    fixture: fixture,
    onComponentRef: onComponentRef,
    onFixtureUpdate: onFixtureUpdate
  });
};

exports.Loader = Loader;
Loader.propTypes = {
  fixture: (0, _moduleType.createModuleType)(_propTypes.object).isRequired,
  proxies: (0, _propTypes.arrayOf)((0, _moduleType.createModuleType)(_propTypes.func)),
  onFixtureUpdate: _propTypes.func,
  onComponentRef: _propTypes.func
};
Loader.defaultProps = {
  proxies: [],
  onFixtureUpdate: function onFixtureUpdate() {},
  onComponentRef: function onComponentRef() {}
};
