// react-cosmos-fixture/dist/testHelpers/components.js
"use strict";
/* tslint:disable:max-classes-per-file */
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
exports.__esModule = true;
var React = require("react");
function HelloMessage(_a) {
    var name = _a.name;
    return React.createElement(React.Fragment, null, "Hello " + (name || 'Stranger'));
}
exports.HelloMessage = HelloMessage;
function Wrapper(_a) {
    var children = _a.children;
    return React.createElement(React.Fragment, null, "children");
}
exports.Wrapper = Wrapper;
// This one is meant for attaching refs
var HelloMessageCls = /** @class */ (function (_super) {
    __extends(HelloMessageCls, _super);
    function HelloMessageCls() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    HelloMessageCls.prototype.render = function () {
        return "Hello " + (this.props.name || 'Stranger');
    };
    return HelloMessageCls;
}(React.Component));
exports.HelloMessageCls = HelloMessageCls;
var Counter = /** @class */ (function (_super) {
    __extends(Counter, _super);
    function Counter() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.state = { count: 0 };
        return _this;
    }
    Counter.prototype.render = function () {
        var count = this.state.count;
        return typeof count === 'number' ? count + " times" : 'Missing count';
    };
    return Counter;
}(React.Component));
exports.Counter = Counter;
var CoolCounter = /** @class */ (function (_super) {
    __extends(CoolCounter, _super);
    function CoolCounter() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.state = { count: 0 };
        return _this;
    }
    CoolCounter.prototype.render = function () {
        return this.state.count + " timez";
    };
    return CoolCounter;
}(React.Component));
exports.CoolCounter = CoolCounter;
var SuffixCounter = /** @class */ (function (_super) {
    __extends(SuffixCounter, _super);
    function SuffixCounter() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.state = { count: 0 };
        return _this;
    }
    SuffixCounter.prototype.render = function () {
        return this.state.count + " " + this.props.suffix;
    };
    return SuffixCounter;
}(React.Component));
exports.SuffixCounter = SuffixCounter;
