// react-cosmos-loader/dist/components/ErrorCatchProxy/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime-corejs2/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime-corejs2/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createErrorCatchProxy = createErrorCatchProxy;

var _extends2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/extends"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/possibleConstructorReturn"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/inherits"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/assertThisInitialized"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _styles = require("./styles");

// Returning a class creator to be consistent with the other proxies and to be
// able to add options in the future without breaking API
function createErrorCatchProxy() {
  var ErrorCatchProxy =
  /*#__PURE__*/
  function (_Component) {
    (0, _inherits2.default)(ErrorCatchProxy, _Component);

    function ErrorCatchProxy() {
      var _getPrototypeOf2;

      var _this;

      (0, _classCallCheck2.default)(this, ErrorCatchProxy);

      for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
        args[_key] = arguments[_key];
      }

      _this = (0, _possibleConstructorReturn2.default)(this, (_getPrototypeOf2 = (0, _getPrototypeOf3.default)(ErrorCatchProxy)).call.apply(_getPrototypeOf2, [this].concat(args)));
      (0, _defineProperty2.default)((0, _assertThisInitialized2.default)((0, _assertThisInitialized2.default)(_this)), "state", {
        hasError: false,
        errorMessage: ''
      });
      return _this;
    }

    (0, _createClass2.default)(ErrorCatchProxy, [{
      key: "componentDidCatch",
      value: function componentDidCatch(error, info) {
        console.log(error, info);
        this.setState({
          hasError: true,
          errorMessage: error.message
        });
      }
    }, {
      key: "render",
      value: function render() {
        return this.state.hasError ? this.renderError() : this.renderNextProxy();
      }
    }, {
      key: "renderError",
      value: function renderError() {
        return _react.default.createElement("div", {
          style: _styles.rootStyles
        }, _react.default.createElement("div", null, _react.default.createElement("strong", null, "Ouch, something wrong!")), _react.default.createElement("div", null, _react.default.createElement("pre", {
          style: _styles.codeStyles
        }, this.state.errorMessage)), _react.default.createElement("div", null, "Check console for more info."));
      }
    }, {
      key: "renderNextProxy",
      value: function renderNextProxy() {
        var nextProxy = this.props.nextProxy;
        return _react.default.createElement(nextProxy.value, (0, _extends2.default)({}, this.props, {
          nextProxy: nextProxy.next()
        }));
      }
    }]);
    return ErrorCatchProxy;
  }(_react.Component);

  return ErrorCatchProxy;
}
