// react-cosmos/dist/server/shared/config-next.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getPlaygroundConfig = getPlaygroundConfig;
exports.FIXTURE_FILE_SUFFIX = exports.FIXTURES_DIR = void 0;
// TODO: Convert constants from this file into user config. But keep these
// values hardcoded in the beta-testing period. This simplifies development and
// promotes designing of good defaults.
// Related https://github.com/react-cosmos/react-cosmos/issues/488
var FIXTURES_DIR = '__jsxfixtures__';
exports.FIXTURES_DIR = FIXTURES_DIR;
var FIXTURE_FILE_SUFFIX = 'jsxfixture';
exports.FIXTURE_FILE_SUFFIX = FIXTURE_FILE_SUFFIX;

function getPlaygroundConfig(_ref) {
  var playgroundOpts = _ref.playgroundOpts,
      devServerOn = _ref.devServerOn;
  return {
    core: {
      projectId: playgroundOpts.projectKey,
      fixturesDir: FIXTURES_DIR,
      fixtureFileSuffix: FIXTURE_FILE_SUFFIX,
      devServerOn: devServerOn,
      webRendererUrl: playgroundOpts.platform === 'web' ? playgroundOpts.loaderUri : null
    }
  };
}
